New invitation flow - part 3

gowsik-ragunath opened the pull request:
## Notion card

## Summary
<!-- Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change. -->

- Added new invitation flow
- Added policy, controller, and service test cases

### Flow:

- Owner or admin can invite users from the team member page
- If the user is a new user then they will receive an email with an invitation token, on click user will be added as a member and it will take them to the password reset page.
- If the user already exists in Miru then that user will receive an email with an invitation token, on click they will be added as a team member then if they are logged in they will be taken to the root page, if not then they will be redirected to sign in page.

## Preview
<!-- Please include a short loom video previewing the changes made in the PR. This will help
the reviewers visualize and get context at a glance -->

https://www.loom.com/share/478ef54739c54b85a5abcfe2e54cf3cf

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. Please also list any relevant details for your
test configuration -->

### Checklist:

- [X] I have manually tested all workflows
- [X] I have performed a self-review of my own code
- [X] I have added automated tests for my code

keshavbiswa commented on app/controllers/users/invitations_controller.rb:
```diff
@@ -0,0 +1,87 @@
+    redirect_to team_index_path
+  end
+
+  def edit
+    authorize current_user, policy_class: InvitationPolicy
+  end
+
+  def update
+    authorize current_user, policy_class: InvitationPolicy
+
+    if @invitation.update(invitation_params)
+      flash[:success] = t(".success")
+    else
+      set_error_flash
+    end
+
+    redirect_to team_index_path
+  end
```
Should we be able to update existing invitations?

gowsik-ragunath replied:
I have added validation to prevent updating an invitation's `recipient_email` field and in frontend made the email field to be read-only. Only `first name` and `last name` fields can be updated from the edit invitation form.

keshavbiswa commented on app/controllers/users/invitations_controller.rb:
```diff
@@ -0,0 +1,87 @@
+  end
+
+  def edit
+    authorize current_user, policy_class: InvitationPolicy
```
Shouldn't this be authorize @invitation ?

gowsik-ragunath replied:
Replaced with `@invitation`

apoorv1316 commented:
@gowsik-ragunath When we invite a user we get an email that has a link. I did something and the link expired eventually and I received the expired link toast too when I clicked on the link, but the user was added to the workspace. So I think it's not right. Also when I tried deleting that user and reinviting him again I received the error shown in the image 
<img width="1792" alt="Screenshot 2022-07-07 at 8 51 50 AM" src="https://user-images.githubusercontent.com/18750194/177684252-4c5670e8-585b-454f-a9a4-6d1fe8bca2b0.png">

gowsik-ragunath commented:
@keshavbiswa I have removed the role memorizing method `admin_or_owner?` as you mentioned in this #534

gowsik-ragunath commented:
@apoorv1316 I have fixed the test cases that failed in the CI test run.

Also, I have fixed the issue in re-inviting a team member again to the workspace but this re-invited user after accepting the invite won't be able to access the workspace until [this issue](https://www.notion.so/saeloun/Bug-in-user-invitation-flow-40ca07827ec54df9be04d4dfe329460b) is fixed which I think Harshaan is working on.

supriya3105 commented:
@keshavbiswa @akhilgkrishnan Please review the PR and approve if all review comments have been incorporated. @apoorv1316 Please test this.

keshavbiswa commented on app/controllers/users/invitations_controller.rb:
```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+class Users::InvitationsController < ApplicationController
```
Any reason why this is Users::InvitationsController and not InvitationsController directly?

gowsik-ragunath replied:
No specific reason I just used the same controller folder structure devise invitation followed. will move this controller out of `Users` folder.

gowsik-ragunath replied:
Moved this logic to `InvitationsController`.

keshavbiswa commented on app/controllers/users/invitations_controller.rb:
```diff
@@ -0,0 +1,87 @@
+    @invitation.sender = current_user
+
+    if @invitation.save
+      flash[:success] = t(".success")
```
```suggestion
      redirect_to team_index_path, success:  t(".success")
```
Change in other places too.

gowsik-ragunath replied:
https://github.com/saeloun/miru-web/blob/5c20577e08e2ca99a2946c479e5672aa3b4894a2/app/controllers/users/invitations_controller.rb#L15-L21

We are using a custom method to set error flash messages and redirect users to the same path.

```
 if @invitation.save 
    redirect_to team_index_path, success:  t(".success")
 else 
   set_error_flash 
   redirect_to team_index_path
 end
```

Doing this will be kind of redundant.

keshavbiswa commented on app/controllers/users/invitations_controller.rb:
```diff
@@ -0,0 +1,87 @@
+    redirect_to team_index_path
+  end
+
+  def accept
```
Let's follow REST API standard and have controllers for this type of action. Invitations::Accept#index sounds good.

gowsik-ragunath replied:
Yeah sure, but do we need to create two controllers for invitations?

> Invitations::Accept#index sounds good.

```
- Invitations(directory)
 |   
 |_ Accept controller
 |
 |_ Invite controller
```

Accept controller will be solely used for accepting the invitations and the invite controller will be used for creating, updating, and deleting the invitations.

keshavbiswa replied:
Invitations#create can be used for creating the invitations, and Invitations::Accept#create can be used for accepting the invitte
Similarly, Invitations::Decline#create can be used for declining the invitation.

gowsik-ragunath replied:
Created separate controllers.

keshavbiswa commented on app/models/user.rb:
```diff
@@ -13,6 +13,13 @@
 #  email                  :string           default(""), not null
 #  encrypted_password     :string           default(""), not null
 #  first_name             :string           not null
+#  invitation_accepted_at :datetime
+#  invitation_created_at  :datetime
+#  invitation_limit       :integer
+#  invitation_sent_at     :datetime
+#  invitation_token       :string
+#  invitations_count      :integer          default(0)
+#  invited_by_type        :string
```
I thought we deleted these attributes? Did it get commited by mistaek?

gowsik-ragunath replied:
Yeah we have removed it but I think switch between other branches and running migration might have re-added this again, will remove it.

gowsik-ragunath replied:
Removed devise invitable related fields

keshavbiswa commented on app/services/create_invited_user_service.rb:
```diff
@@ -0,0 +1,83 @@
+    end
+
+    def find_or_create_user!
+      if User.exists?(email: invitation.recipient_email)
+        @user = User.find_by!(email: invitation.recipient_email)
```
```suggestion
      if  @user = User.find_by(email: invitation.recipient_email)
```

gowsik-ragunath replied:
Removed ! from find_by

keshavbiswa replied:
No not just that
The above two lines can be merged into one.

gowsik-ragunath replied:
Merged into one line.

keshavbiswa commented on app/services/create_invited_user_service.rb:
```diff
@@ -0,0 +1,83 @@
+        @user.update!(current_workspace_id: invitation.company.id)
+      else
+        create_invited_user!
+        create_reset_password_token
```
I'm not sure why we're creating reset_password_token? I think we discussed we won't be redirecting users to reset password page.

gowsik-ragunath replied:
We are not sending any email to reset their account password, we are generating this token to take them to the set password page, once they set the password they can login to miru.

https://saelounus.slack.com/archives/C033ZCDGJ10/p1656417801266659?thread_ts=1656415661.049029&cid=C033ZCDGJ10

keshavbiswa commented on app/views/internal_api/v1/team/index.json.jbuilder:
```diff
@@ -4,14 +4,14 @@ json.key_format! camelize: :lower
 
 def team_member_status(member)
-  return unless current_user.admin_or_owner?(current_company) &&
+  return unless current_user.has_role?(:owner, current_company) || current_user.has_role?(:admin, current_company) &&
```
This will fire two queries to check role, look up has_any_role? documentation in rolify to do this in one query.

gowsik-ragunath replied:
Oh yes, Instead of removing the `admin_or_owner?` memorizer method I removed the whole fix, will re-add.

gowsik-ragunath replied:
Replaced with `has_any_role?`

keshavbiswa commented on app/views/team/_invitation.html.erb:
```diff
@@ -1,43 +1,41 @@
-<td id="pending-<%= member.id %>" class="table__data">
-  <% if current_user.admin_or_owner?(current_company) %>
+<td id="pending-<%= invitation.id %>" class="table__data">
+  <% if current_user.has_role?(:owner, current_company) || current_user.has_role?(:admin, current_company) %>
```
Same concerns as above.

gowsik-ragunath replied:
Replaced with `has_any_role?`

supriya3105 commented:
@keshavbiswa Can you review this PR again as Gowsik have responded to your earlier review comments.

keshavbiswa commented on app/controllers/invitations_controller.rb:
```diff
@@ -0,0 +1,69 @@
+    @invitation.sender = current_user
+
+    if @invitation.save
+      flash[:success] = t(".success")
```
```suggestion
      redirect_to team_index_path, success: t(".success")
```

gowsik-ragunath replied:
Updated

keshavbiswa commented on app/controllers/invitations_controller.rb:
```diff
@@ -0,0 +1,69 @@
+    if @invitation.save
+      flash[:success] = t(".success")
+    else
+      set_error_flash
```
```suggestion
      redirect_to team_index_path, error: set_error_flash
```

gowsik-ragunath replied:
Updated

keshavbiswa commented on app/models/invitation.rb:
```diff
@@ -23,26 +23,82 @@
+  # Callbacks
+  before_validation :set_token, on: :create
+  before_validation :set_expired_at, on: :create
+  after_create :send_invitation_mail
```
Let's use after_create_commit and use sidekiq to process the email, so that we don't block the invitation creation.

gowsik-ragunath replied:
Changed to `after_create_commit`, we are already processing this email in sidekiq we are using `deliver_later` action mailer method to enqueue this mail in active job queue.

keshavbiswa commented on app/services/create_invited_user_service.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+
+    def find_or_create_user!
+      if (@user = User.find_by(email: invitation.recipient_email)).present?
```
I don't think we need `present?` check. Could you confirm that once?

gowsik-ragunath replied:
Yeah, `present?` is not needed but when I had a direct assignment robocop throws this warning `found = in conditional, should be ==` and prevented the code from committing.

[Closed issue ref.](https://github.com/rubocop/ruby-style-guide/issues/53)

In this issue, they have mentioned that adding parenthesis is enough, will try this and update the code.

gowsik-ragunath replied:
Removed `present?`.

apoorv1316 commented:
@gowsik-ragunath I am still facing the same issue which I was facing earlier. You said you will fix the deletion of users and reinviting them again. Can you please do that?

gowsik-ragunath commented:
@apoorv1316 I have fixed the bug in displaying the correct role of the user in the current workspace.

<img width="1208" alt="Screenshot 2022-07-12 at 2 12 59 PM" src="https://user-images.githubusercontent.com/52308930/178449275-a3a20b5c-857a-453b-9dfc-bba142b63f94.png">

supriya3105 commented:
@keshavbiswa As discussed on stand up call , please review again and approve if no major comments. Any minor comments can be added as separate task.
